src/transform/src/normalize_lets.rs

```diff
@@ -0,0 +1,360 @@
+        while let Some((id, mut expr)) = worklist.pop() {
+            digest_lets_helper(&mut expr, &mut worklist);
+            lets.insert(id, expr);
+        }
+        lets
+    }
+
+    /// Extract all `Let` bindings from `expr`, into `(id, value)` pairs.
+    ///
+    /// Importantly, the `value` pairs may not be `Let`-free, and must be further processed.
+    fn digest_lets_helper(
+        expr: &mut MirRelationExpr,
+        bindings: &mut Vec<(LocalId, MirRelationExpr)>,
+    ) {
+        let mut to_visit = vec![expr];
+        while let Some(expr) = to_visit.pop() {
+            if let MirRelationExpr::Let { id, value, body } = expr {
+                bindings.push((*id, value.take_dangerous()));
+                *expr = body.take_dangerous();
+                to_visit.push(expr);
+            } else {
+                to_visit.extend(expr.children_mut());
+            }
+        }
+    }
+
+    /// Substitute `Get{id}` expressions for any proposed expressions.
+    ///
+    /// The proposed expressions can be proposed either to be taken or cloned.
+    pub(super) fn inline_gets(
+        expr: &mut MirRelationExpr,
+        to_take: &mut BTreeMap<LocalId, Option<MirRelationExpr>>,
+        to_clone: &BTreeMap<LocalId, MirRelationExpr>,
+    ) {
+        let mut worklist = vec![expr];
+        while let Some(expr) = worklist.pop() {
+            if let MirRelationExpr::Get {
+                id: Id::Local(id), ..
+            } = expr
+            {
+                if let Some(value) = to_take.get_mut(id) {
+                    if let Some(value) = value.take() {
+                        *expr = value;
+                        worklist.push(expr);
+                    } else {
+                        panic!("Value already taken for {:?}", id);
+                    }
+                } else if let Some(value) = to_clone.get(id) {
+                    *expr = value.clone();
+                    worklist.push(expr);
+                }
+            } else {
+                worklist.extend(expr.children_mut());
+            }
+        }
+    }
+
+    pub(super) fn renumber_bindings_helper(
+        guard: &RecursionGuard,
+        relation: &mut MirRelationExpr,
+        remap: &mut BTreeMap<LocalId, LocalId>,
+        id_gen: &mut IdGen,
+    ) -> Result<(), crate::TransformError> {
+        guard.checked_recur(|_| {
+            match relation {
+                MirRelationExpr::Let { id, value, body } => {
+                    renumber_bindings_helper(guard, value, remap, id_gen)?;
+                    // If a local id, assign a new identifier and refresh the type.
+                    let new_id = LocalId::new(id_gen.allocate_id());
+                    let prev = remap.insert(id.clone(), new_id);
+                    renumber_bindings_helper(guard, body, remap, id_gen)?;
+                    remap.remove(id);
+                    if let Some(prev_stuff) = prev {
+                        remap.insert(id.clone(), prev_stuff);
```

Comment:
The main source that I'm aware of is `lowering.rs`, which includes defensive measures around this, which made me think that there are reasons in there that this might happen. I'm not sure it is an error if it happens, though. I don't mind putting a `warn` in but I'm also not sure there is really a problem. I'd be delighted to avoid it though, and let any occurrence of MIR shadowing be fundamentally incorrect.

Reply:
It's not fundamentally incorrect, but for me it falls under the "A recoverable error that is still unexpected" `WARN` case from the guidelines. The code will still work, but hopefully this will highlight cases where we violate the assumption that such trees should not exist.